### PR TITLE
gh-102013: Add PyUnstable_GC_VisitObjects

### PR DESCRIPTION
**Pull request by jbower-fb**
* Issue: gh-102013

### CODE REVIEW

File: Misc/NEWS.d/next/C API/2023-02-18-00-55-14.gh-issue-102013.83mrtI.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Add a new C-API function for iterating over GC'able objects using a callback: ``PyGC_VisitObjects``.
```
Comment (erlend-aasland) [inline Misc/NEWS.d/next/C API/2023-02-18-00-55-14.gh-issue-102013.83mrtI.rst]: You forgot the `PyUnstable_` prefix ;)

File: Modules/_testcapimodule.c [hunk 2]
```diff
@@ -3517,6 +3517,76 @@ err_restore(PyObject *self, PyObject *args) {
     return NULL;
 }
 
+struct gc_visit_state_basic {
+    PyObject *target;
+    int found;
+};
+
+static int
+gc_visit_callback_basic(PyObject *obj, void* arg) {
```
Comment (erlend-aasland) [inline Modules/_testcapimodule.c]: PEP-7

```suggestion
gc_visit_callback_basic(PyObject *obj, void *arg)
{
```

File: Modules/_testcapimodule.c [hunk 3]
```diff
@@ -3517,6 +3517,76 @@ err_restore(PyObject *self, PyObject *args) {
...
+    struct gc_visit_state_basic *state = (struct gc_visit_state_basic *)arg;
+    if (obj == state->target) {
+        state->found = 1;
+        return 0;
+    }
+    return 1;
+}
+
+static PyObject *
+test_gc_visit_objects_basic(PyObject *Py_UNUSED(self), PyObject *Py_UNUSED(ignored)) {
```
Comment (erlend-aasland) [inline Modules/_testcapimodule.c]: PEP-7

```suggestion
test_gc_visit_objects_basic(PyObject *Py_UNUSED(self), PyObject *Py_UNUSED(ignored))
{
```

File: Modules/_testcapimodule.c [hunk 4]
```diff
@@ -3517,6 +3517,76 @@ err_restore(PyObject *self, PyObject *args) {
...
+    int error = 0;
+    goto done;
+error:
+    error = 1;
+done:
+    Py_XDECREF(obj);
+    if (error) {
+        return NULL;
+    }
+    Py_RETURN_NONE;
```
Comment (erlend-aasland) [inline Modules/_testcapimodule.c]: This looks a little convoluted with all the labels. What about this simplification?

```suggestion
    obj = PyList_New(0);
    if (obj == NULL) {
        return NULL;
    }
    state.target = obj;
    state.found = 0;
    
    PyUnstable_GC_VisitObjects(gc_visit_callback_basic, &state);
    Py_DECREF(obj);
    if (!state.found) {
        PyErr_SetString(
             PyExc_AssertionError,
             "test_gc_visit_objects_basic: Didn't find live list");
         return NULL;
    }
    Py_RETURN_NONE;
```

File: Modules/_testcapimodule.c [hunk 5]
```diff
@@ -3517,6 +3517,76 @@ err_restore(PyObject *self, PyObject *args) {
...
+}
+
+static int
+gc_visit_callback_exit_early(PyObject *obj, void* arg) {
```
Comment (erlend-aasland) [inline Modules/_testcapimodule.c]: PEP-7

```suggestion
gc_visit_callback_exit_early(PyObject *obj, void *arg)
 {
```

File: Modules/_testcapimodule.c [hunk 6]
```diff
@@ -3517,6 +3517,76 @@ err_restore(PyObject *self, PyObject *args) {
...
+    int *visited_i = (int *)arg;
+    (*visited_i)++;
+    if (*visited_i == 2) {
+        return 0;
+    }
+    return 1;
+}
+
+static PyObject *
+test_gc_visit_objects_exit_early(PyObject *Py_UNUSED(self), PyObject *Py_UNUSED(ignored)) {
```
Comment (erlend-aasland) [inline Modules/_testcapimodule.c]: PEP-7

```suggestion
test_gc_visit_objects_exit_early(PyObject *Py_UNUSED(self), PyObject *Py_UNUSED(ignored))
{
```

File: Modules/gcmodule.c [hunk 7]
```diff
@@ -2401,3 +2401,26 @@ PyObject_GC_IsFinalized(PyObject *obj)
     }
     return 0;
 }
+
+void
+PyUnstable_GC_VisitObjects(gcvisitobjects_t callback, void *arg) {
```
Comment (erlend-aasland) [inline Modules/gcmodule.c]:
```suggestion
PyUnstable_GC_VisitObjects(gcvisitobjects_t callback, void *arg)
{
```

File: Modules/gcmodule.c [hunk 8]
```diff
@@ -2401,3 +2401,26 @@ PyObject_GC_IsFinalized(PyObject *obj)
...
+            PyObject *op = FROM_GC(gc);
+            Py_INCREF(op);
+            int res = callback(op, arg);
+            Py_DECREF(op);
+            if (!res) {
+              goto done;
+            }
+        }
+    }
+done:;
```
Comment (erlend-aasland) [inline Modules/gcmodule.c]: There's an extra semi-colon here.

File: Modules/gcmodule.c [hunk 9]
```diff
@@ -2401,3 +2401,26 @@ PyObject_GC_IsFinalized(PyObject *obj)
     }
     return 0;
 }
+
+void
+PyUnstable_GC_VisitObjects(gcvisitobjects_t callback, void *arg) {
+    size_t i;
+    GCState* gcstate = get_gc_state();
```
Comment (erlend-aasland) [inline Modules/gcmodule.c]:
```suggestion
    GCState *gcstate = get_gc_state();
```

File: Modules/gcmodule.c [hunk 10]
```diff
@@ -2401,3 +2401,26 @@ PyObject_GC_IsFinalized(PyObject *obj)
...
+    for (i = 0; i < NUM_GENERATIONS; i++) {
+        PyGC_Head *gc_list, *gc;
+        gc_list = GEN_HEAD(gcstate, i);
+        for (gc = GC_NEXT(gc_list); gc != gc_list; gc = GC_NEXT(gc)) {
+            PyObject *op = FROM_GC(gc);
+            Py_INCREF(op);
+            int res = callback(op, arg);
+            Py_DECREF(op);
+            if (!res) {
+              goto done;
```
Comment (erlend-aasland) [inline Modules/gcmodule.c]: 4 spaces indent (PEP-7)

```suggestion
                goto done;
```

File: Include/objimpl.h [hunk 11]
```diff
@@ -157,6 +157,25 @@ PyAPI_FUNC(int) PyGC_Enable(void);
...
+ * to the supplied arg. Returning 0 from the callback ends iteration, returning
+ * 1 allows iteration to continue. Returning any other value may result in
+ * undefined behaviour.
+ *
+ * If new objects are (de)allocated by the callback it is undefined if they
+ * will be visited.
+
+ * Garbage collection is disabled during operation. Explicitly running a
+ * collection in the callback may lead to undefined behaviour e.g. visiting the
+ * same objects multiple times or not at all.
```
Comment (erlend-aasland) [inline Include/objimpl.h:173]: According to the PEP-689 spec, unstable APIs must be documented, so IMO this comment should be added to `Doc/c-api/gcsupport.rst`.

Comment (jbower-fb) [reply Include/objimpl.h:173]: I wasn't sure where to add this, thanks for the pointer.

**Comment by jbower-fb** [comment]
Thank you for the feedback @erlend-aasland. I'll try to get better at PEP-7 compliance in future!

**Comment by jbower-fb** [comment]
Hey @pablogsal, any thoughts on this?

File: Include/objimpl.h [hunk 12]
```diff
@@ -157,6 +157,25 @@ PyAPI_FUNC(int) PyGC_Enable(void);
 PyAPI_FUNC(int) PyGC_Disable(void);
 PyAPI_FUNC(int) PyGC_IsEnabled(void);
 
+
+#if !defined(Py_LIMITED_API)
```
Comment (encukou) [inline Include/objimpl.h:161]: You can put this in `Include/cpython/objimpl.h`, which is only included for non-limited API. But that's not a blocker.

**Review by encukou — APPROVED** [review]
API-wise, this looks good!

**Comment by pablogsal** [comment]
> Hey @pablogsal, any thoughts on this?

Will make a review later this week. Apologies for the delay but I am a bit swamped recently

File: Doc/c-api/gcsupport.rst [hunk 13]
```diff
@@ -228,3 +228,36 @@ garbage collection runs.
    Returns the current state, 0 for disabled and 1 for enabled.
 
    .. versionadded:: 3.10
+
+
+Querying Garbage Collector State
+--------------------------------
+
+The C-API provides the following inerface for querying information about
```
Comment (erlend-aasland) [inline Doc/c-api/gcsupport.rst]:
```suggestion
The C-API provides the following interface for querying information about
```

File: Doc/c-api/gcsupport.rst [hunk 14]
```diff
@@ -228,3 +228,36 @@ garbage collection runs.
...
+   Run supplied callback on all live GC-capable objects.
+
+   .. warning::
+      If new objects are (de)allocated by the callback it is undefined if they
+      will be visited.
+
+   .. warning::
+      Garbage collection is disabled during operation. Explicitly running a collection
+      in the callback may lead to undefined behaviour e.g. visiting the same objects
+      multiple times or not at all.
```
Comment (erlend-aasland) [inline Doc/c-api/gcsupport.rst]: Would it be ok to merge these two warning boxes, for less visual clutter?

File: Doc/c-api/gcsupport.rst [hunk 15]
```diff
@@ -228,3 +228,36 @@ garbage collection runs.
...
+
+Querying Garbage Collector State
+--------------------------------
+
+The C-API provides the following inerface for querying information about
+the garbage collector.
+
+.. c:function:: void PyUnstable_GC_VisitObjects(gcvisitobjects_t callback, void *arg)
+
+   Run supplied callback on all live GC-capable objects.
```
Comment (erlend-aasland) [inline Doc/c-api/gcsupport.rst]: Format parameters correctly by wrapping them with a pair of \* (as per [devguide recommendations](https://devguide.python.org/documentation/markup/#quick-reference))
Also note that *arg* is passed to the callback.
(I made a clumsy suggestion to improve this; feel free to improve my suggestion. Perhaps a new sentence is better.)

```suggestion
   Run *callback*, with *arg* passed, on all live GC-capable objects.
```

File: Doc/c-api/gcsupport.rst [hunk 16]
```diff
@@ -228,3 +228,36 @@ garbage collection runs.
...
+      multiple times or not at all.
+
+   .. versionadded:: 3.12
+
+.. c:type:: int (*gcvisitobjects_t)(PyObject *object, void *arg)
+
+   Type of the visitor function to be passed to :c:func:`PyUnstable_GC_VisitObjects`.
+   ``arg`` is the same as passed to ``PyUnstable_GC_VisitObjects``. Returning ``0``
+   allows iteration to continue, returning ``1`` stops further iteration. Returning any
+   other value has undefined behavior.
```
Comment (erlend-aasland) [inline Doc/c-api/gcsupport.rst]: - Fix argument format (pair of \*, not backticks)
- Slightly more succinct wording regarding return values

```suggestion
   *arg* is the same as the *arg* passed to ``PyUnstable_GC_VisitObjects``.
   Return ``0`` to continue iteration, return ``1`` to stop iteration.
   Returning any other value is undefined behavior.
```

Perhaps the last sentence should be a `.. warning::`?

File: Modules/_testcapimodule.c [hunk 17]
```diff
@@ -3316,6 +3316,71 @@ function_set_kw_defaults(PyObject *self, PyObject *args)
...
+    struct gc_visit_state_basic *state = (struct gc_visit_state_basic *)arg;
+    if (obj == state->target) {
+        state->found = 1;
+        return 0;
+    }
+    return 1;
+}
+
+static PyObject *
+test_gc_visit_objects_basic(PyObject *Py_UNUSED(self), PyObject *Py_UNUSED(ignored))
```
Comment (erlend-aasland) [inline Modules/_testcapimodule.c]:
```suggestion
test_gc_visit_objects_basic(PyObject *Py_UNUSED(self),
                            PyObject *Py_UNUSED(ignored))
```

File: Modules/_testcapimodule.c [hunk 18]
```diff
@@ -3316,6 +3316,71 @@ function_set_kw_defaults(PyObject *self, PyObject *args)
...
+    int *visited_i = (int *)arg;
+    (*visited_i)++;
+    if (*visited_i == 2) {
+        return 0;
+    }
+    return 1;
+}
+
+static PyObject *
+test_gc_visit_objects_exit_early(PyObject *Py_UNUSED(self), PyObject *Py_UNUSED(ignored))
```
Comment (erlend-aasland) [inline Modules/_testcapimodule.c]:
```suggestion
test_gc_visit_objects_exit_early(PyObject *Py_UNUSED(self),
                                 PyObject *Py_UNUSED(ignored))
```

File: Include/objimpl.h (hunk 12 above)
Comment (jbower-fb) [reply Include/objimpl.h:161]: I'm not very familiar with the conventions here, is that what we should do with public API (even if it's unstable)?

**Review by pablogsal — APPROVED** [review]
Apologies for the delay although the advantage of being late is that everyone has already done an excellent job with the reviews.

LGTM

Thanks a lot for the patience and understanding @jbower-fb and thanks for the great work :)

File: Include/objimpl.h (hunk 12 above)
Comment (encukou) [reply Include/objimpl.h:161]: Yes.

[Details are in the Devguide](https://devguide.python.org/developer-workflow/c-api/index.html).

**Pull request merged**